clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -3307,6 +3307,29 @@ void Sema::CheckShadowInheritedFields(const SourceLocation &Loc,
   }
 }
 
+template <typename AttrType>
+inline static bool HasAttribute(const QualType &T) {
+  if (const TagDecl *TD = T->getAsTagDecl())
+    return TD->hasAttr<AttrType>();
+  if (const TypedefType *TDT = T->getAs<TypedefType>())
+    return TDT->getDecl()->hasAttr<AttrType>();
+  return false;
+}
+
+inline static bool IsUnusedPrivateField(FieldDecl *FD) {
+  if (FD->getAccess() == AS_private && FD->getDeclName()) {
+    QualType FieldType = FD->getType();
+    if (HasAttribute<WarnUnusedAttr>(FieldType))
```

Comment:
Is it okay to proceed with a warning if the `[warn_unused]` attribute is present, or should additional constraints be considered?

nit: would it make sense to move this complex condition to a helper function or better to extend it?

Reply:
I'm not sure which additional constraints to consider?  Can you clarify what you mean?

Reply:
I meant to check whether the declaration is explicit or context-dependent.

Reply:
The FieldDecl?  I guess I can't think of any conditions where it could be context dependent/implicit?  Or am I missing something again?